### ISSUE TITLE
Request for more strict release rules and release only default branch

### ISSUE THREAD

**Issue by tomasbrod** [gridcoin-community/Gridcoin-Research#485]
I politely ask, to prevent future confusion on what is and what is not release ready to establish some release rules. For example establish rule, that a release must be first approved by *collaborators* (Rob and Marco).

As it is cumbersome to checkout latest *tag* with git, I ask to also create and maintain a *branch* that will always point only to a released code. Preferably this should be the master branch, or at least the default branch you get when you clone the repo.

Additionally a branch that is supposed to be run in testnet. 
Note that git branches are really cheap to create and delete.

**Comment by tomasbrod** [comment]
For example the master branch could be made protected, like shown on screen-shot below.
[screen-shot](https://user-images.githubusercontent.com/11995799/28967059-b0d0dd16-7919-11e7-99f5-bf8266d0fd38.png) or [settings page](https://github.com/gridcoin/Gridcoin-Research/settings/branches/master)

**Comment by tomasbrod** [comment]
You can set resease branch to point to latest commit from development branch with this command:
```
git branch -f release development
```

**Comment by denravonska** [comment]
It is not that different from what we have now. I guess the difference would be that the tag and merge to master happens in the same push, right? The intention with master is to always have it either containing only release code or the code that is going to be released when @gridcoin wakes up. Given the roles we have and the timezone differences this was the easiest path and it should be pretty clear. I guess those roles can be redefined though.

While a testnet branch would be nice it also adds additional administration overhead. It's already a bit of a hassle to keep track of the three branches we have. I'd say that anything that is not master is testnet only unless explicitly stated otherwise. For example, staging has to go to production at one point or another, but it's done in a controlled manner.

I do think that we should change the default PR branch to point to development but I'm not sure if it's possible without also changing the default branch of the entire repo. Is it? I like the master protection. We should absolutely add that.

So in short, this is the current setup:
- Master: Always release ready, run on production net.
- Staging: About to be release ready. Start on testnet, move to production net before merge.
- Development. Never release ready, always on testnet.

**Comment by tomasbrod** [comment]
Branches are just a pointer to commit. They can be set to point wherever (careful for non-fast-forward changes).

> Master: Always release ready, run on production net.

I am pleased by this rule. But I don't want to force anything, that branch does not have to be named "master".

Like: Release is prepared in staging branch and when you Admins agree to call it a release it is:
* [optional update changelog]
* [optional create (empty), gpg signed commit]
* create release tag for the latest commit
* update master to point to staging (using git merge or github merge)

**Comment by Lenni** [comment]
I have noticed that we push changes to the code like the V8 changes diretly into "master". Aren't we supposed to firstly use the development branch for this?

**Comment by denravonska** [comment]
@Lenni V8 was merged from PR to the development branched and later merged to staging for testing. After that it was merged to master and released. The subsequent fixes were merged to the closest stable point, which was staging. Since it's a hotfix release we can't take the development->staging->master route as that would bring in untested code.

**Comment by Lenni** [comment]
Oh, I see. Well it was just confusing for a newbie like me too see those changes being immediately pushed into the master branch.

**Comment by ghost** [comment]
@Lenni maybe should have been a branch called hotfix or something then merged into master

**Comment by tomasbrod** [comment]
No need to create a static hotfix branch. You create a branch to fix release when it is needed, then abandon/delete it.

**Comment by tomasbrod** [comment]
But it looks like there is no go with this.
